topology/dynamic.go

```diff
@@ -0,0 +1,222 @@
+
+	if t.closed {
+		return
+	}
+
+	t.closed = true
+
+	t.w.Close()
+	t.s.Close()
+}
+
+func waitOnInit(t xwatch.Source, d time.Duration) error {
+	if d <= 0 {
+		return nil
+	}
+	select {
+	case <-t.WaitInit():
+		return nil
+	case <-time.After(d):
+		return errInitTimeOut
+	}
+}
+
+// serviceTopologyInput implements xwatch.SourceInput with retry
+type serviceTopologyInput struct {
+	w       xwatch.Watch
+	retrier xretry.Retrier
+	hashGen sharding.HashGen
+	closed  bool
+	// NB(chaowang) the watch may be initiated and not get any updates on its channel any time soon,
+	// so we should try whatever value the watch has as first poll to avoid source init time out.
+	// If the value the watch has is invalid, the source is still not initiated
+	poll int
+}
+
+func newServiceTopologyInput(w xwatch.Watch, h sharding.HashGen, opts xretry.Options) xwatch.SourceInput {
+	return &serviceTopologyInput{w: w, hashGen: h, retrier: xretry.NewRetrier(opts)}
+}
+
+func (s *serviceTopologyInput) Poll() (interface{}, error) {
+	var m Map
+	if attemptErr := s.retrier.Attempt(func() error {
+		if s.poll > 0 {
+			// for normal attempts (from 2nd attempt on), block on the notification channel
+			if _, ok := <-s.w.C(); !ok {
+				s.closed = true
+				return xwatch.ErrSourceClosed
+			}
+		}
+		s.poll++
+
+		var err error
+		m, err = newMapFromServiceInstances(s.w.Get(), s.hashGen)
+		return err
+	}); attemptErr != nil {
+		return nil, attemptErr
+	}
+
+	return m, nil
+}
+
+func newMapFromServiceInstances(data interface{}, hashGen sharding.HashGen) (Map, error) {
+	si, ok := data.([]services.ServiceInstance)
+	if !ok {
+		return nil, errInvalidTopology
+	}
+	to, err := newMapOptionsFromServiceInstances(si, hashGen)
+	if err != nil {
+		return nil, err
+	}
+	if err := to.Validate(); err != nil {
+		return nil, err
+	}
+	return newStaticMap(to), nil
+}
+
+func newMapOptionsFromServiceInstances(instances []services.ServiceInstance, hashGen sharding.HashGen) (StaticOptions, error) {
+	allShards, r := uniqueShardsAndReplicas(instances)
+
+	fn := hashGen(uint32(len(allShards)))
+
+	allShardSet, err := sharding.NewShardSet(allShards, fn)
+	if err != nil {
+		return nil, err
+	}
+
+	hostShardSets := make([]HostShardSet, len(instances))
+	for i, instance := range instances {
+		hs, err := newHostShardSetFromServiceInstance(instance, fn)
+		if err != nil {
+			return nil, err
+		}
+		hostShardSets[i] = hs
+	}
+
+	return NewStaticOptions().
+		Replicas(r).
+		ShardSet(allShardSet).
+		HostShardSets(hostShardSets), nil
+}
+
+func uniqueShardsAndReplicas(instances []services.ServiceInstance) ([]uint32, int) {
+	m := make(map[uint32]int)
+	for _, i := range instances {
+		for _, s := range i.Shards().Shards() {
+			m[s.ID()] = m[s.ID()] + 1
+		}
+	}
+	r := math.MaxUint32
+	s := make([]uint32, 0, len(m))
+
+	// NB(chaowang): We allow different replicas for a shard in case cluster is changing capacity
+	for k, v := range m {
+		s = append(s, k)
+		if r > v {
+			r = v
```

Comment:
Ok so yeah I believe to `Services` interface in `m3cluster` we're going to have to add some kind of queryable endpoint to query for this type of metadata:

```go
type Services interface {
  Advertise(ad Advertisement) error
  Unadvertise(service, id string) error
  QueryInstances(service string, opts QueryOptions) ([]ServiceInstance, error)
  WatchInstances(service string, opts QueryOptions) (xwatch.Watch, error)

  // QueryShardingInfo queries information regarding a service's sharding
  QueryShardingInfo(service string) (shard.ShardingInfo, error)
}
```

Then in the `shard` package:

```go
type ShardingInfo interface {
  Replicas() int
  ShardRange() uint32
}
```

That way you can verify the actual data returned here when setting the assigned shards that all shards meet the baseline replicas that's supposed to be set.  

Also it means when you're creating the `HashFn` you don't have to just get the length of the shards and can do it more explicitly:

```go
func newMapOptionsFromServiceInstances(..., shardingInfo m3clustershard.ShardingInfo, newHashFn sharding.NewHashFnFromShardRange, ...) (StaticOptions, error) {
  // ....
  fn := newHashFn(shardingInfo.ShardRange())
  // ...
}
```